File: 20230620_Tuesday/20230620.rb
# Leetcode: 2090. K Radius Subarray Averages.
# https://leetcode.com/problems/k-radius-subarray-averages/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 659 ms, faster than 75.00% of Ruby online submissions for K Radius
# Subarray Averages.
# Memory Usage: 233.6 MB, less than 25.00% of Ruby online submissions for K
# Radius Subarray Averages.
# 2023.06.20 Daily Challenge.
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer[]}

def get_averages(nums, k)
  # 0 1 2 3 4 5 6 7: k = 2
  # 1 2 i 1 2
  #     k
  answer = Array.new(nums.size, -1)
  return answer if nums.size < k*2 + 1
  # Step one generate sum.
  sum = nums[0...(k*2 + 1)].sum 
  amount = 2*k+1
  answer[k] = (sum/amount)
  return answer if nums.size == k*2 + 1
  (k+1...nums.size - k).each do |i|
    sum = sum - nums[i-k-1] + nums[i+k]
    answer[i] = (sum/amount)
  end
  return answer
end

File: 20230620_Tuesday/20230620.go
// Leetcode: 2090. K Radius Subarray Averages.
// https://leetcode.com/problems/k-radius-subarray-averages/
// = = = = = = = = = = = = = =
// Accpeted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 265 ms, faster than 34.18% of Go online submissions for K Radius
// Subarray Averages.
// Memory Usage: 11.5 MB, less than 25.32% of Go online submissions for K Radius
// Subarray Averages.
// 2023.06.20 Daily Challenge.

package main

import ()

func getAverages(nums []int, k int) []int {
	n, m := len(nums), k*2+1
	answer := make([]int, n)
	// Fill answer with -1;
	for i, _ := range answer {
		answer[i] = -1
	}
	// Return if an interval is bigger then the size.
	if n < m {
		return answer
	}
	// Calculage sum of first inteval.
	sum := 0
	for i := 0; i < m; i += 1 {
		sum += nums[i]
	}
	// Save avg in the middle of the first interval.
	answer[k] = (sum / m)
	// If all array size is equal to inteveral.
	if n == m {
		return answer
	}
	// now loop to fill the middles of all possible interval except the first
	// one.
	for i := k + 1; i < n-k; i += 1 {
		sum = sum - nums[i-k-1] + nums[i+k]
		answer[i] = (sum / m)
	}
	// Eventually return the answer.
	return answer
}
